File: src/main/kotlin/com/perfscope/db/DatabaseLoader.java
package com.perfscope.db;

import com.perfscope.model.CallTreeData;
import com.perfscope.model.CommandData;
import com.perfscope.model.tables.Calls;
import com.perfscope.model.tables.CommThreads;
import com.perfscope.model.tables.Comms;
import com.perfscope.model.tables.Threads;
import com.perfscope.model.tables.records.CommsRecord;
import org.jooq.DSLContext;
import org.jooq.Record3;
import org.jooq.Result;
import org.jooq.impl.DSL;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

public class DatabaseLoader {
    
    private static final Logger logger = LoggerFactory.getLogger(DatabaseLoader.class);
    private static final long ROOT_PARENT_CALL_PATH_ID = 1L;

    public List<CommandData> loadCommands(String databasePath) throws SQLException {
        logger.info("Loading comms with calls from: {}", databasePath);
        List<CommandData> result = new ArrayList<>();
        
        try (DatabaseConnectionHolder dbConnection = new DatabaseConnectionHolder(databasePath)) {
            DSLContext context = dbConnection.getContext();
            
            List<CommsRecord> commsWithCalls = context
                .selectFrom(Comms.COMMS)
                .where(Comms.COMMS.HAS_CALLS.eq(true))
                .fetch();
            
            logger.info("Found {} comms with calls", commsWithCalls.size());
            
            for (CommsRecord comm : commsWithCalls) {
                Result<Record3<Long, Integer, Integer>> threads = context
                    .select(CommThreads.COMM_THREADS.THREAD_ID, Threads.THREADS.PID, Threads.THREADS.TID)
                    .from(CommThreads.COMM_THREADS)
                    .innerJoin(Threads.THREADS).on(CommThreads.COMM_THREADS.THREAD_ID.eq(Threads.THREADS.ID.cast(Long.class)))
                    .where(CommThreads.COMM_THREADS.COMM_ID.eq(comm.getId().longValue()))
                    .fetch();
                
                result.add(new CommandData(comm.getId(), comm.getComm(), threads));
            }
        }
        
        return result;
    }
    
    public Long calculateTotalTimeNanos(String databasePath, Long commId, Long threadId) {
        logger.debug("Calculating total time for comm: {}, thread: {}", commId, threadId);
        try (DatabaseConnectionHolder dbConnection = new DatabaseConnectionHolder(databasePath)) {
            DSLContext queryContext = dbConnection.getContext();

            Result<?> nodes = queryContext
                .select(DSL.sum(Calls.CALLS.RETURN_TIME.minus(Calls.CALLS.CALL_TIME)))
                .from(Calls.CALLS)
                .where(Calls.CALLS.PARENT_CALL_PATH_ID.eq(ROOT_PARENT_CALL_PATH_ID))
                .and(Calls.CALLS.COMM_ID.eq(commId))
                .and(Calls.CALLS.THREAD_ID.eq(threadId))
                .groupBy(Calls.CALLS.CALL_PATH_ID)
                .fetch();
            
            Long totalTimeNanos = 0L;
            for (org.jooq.Record record : nodes) {
                totalTimeNanos += record.get(0, Long.class);
            }
            return totalTimeNanos != 0L ? totalTimeNanos: 1L;
        } catch (Exception e) {
            logger.error("Error calculating total time: {}", e.getMessage(), e);
            return 1L;
        }
    }

    public List<CallTreeData> loadCallTreeNodes(
        String databasePath,
        Long commId,
        Long threadId,
        Long parentCallPathId,
        Long fromCallTime,
        Long toReturnTime) throws SQLException 
    {
        logger.debug("Loading call tree nodes for comm: {}, thread: {}, parent: {}", commId, threadId, parentCallPathId);

        // TODO rewrite this
        try (DatabaseConnectionHolder dbConnection = new DatabaseConnectionHolder(databasePath)) {
            DSLContext queryContext = dbConnection.getContext();

            if (fromCallTime != null) {
                return queryContext.fetch(
                    "SELECT call_path_id, name, short_name, return_time - call_time, " +
                            "insn_count, cyc_count, branch_count, call_time, return_time " +
                            "FROM calls " +
                            "INNER JOIN call_paths ON calls.call_path_id = call_paths.id " +
                            "INNER JOIN symbols ON call_paths.symbol_id = symbols.id " +
                            "INNER JOIN dsos ON symbols.dso_id = dsos.id " +
                            "WHERE parent_call_path_id = ? " +
                            "AND comm_id = ? " +
                            "AND thread_id = ? " +
                            "AND call_time >= ? " +
                            "AND return_time <= ? " +
                            "ORDER BY call_time, call_path_id",
                    parentCallPathId, commId, threadId, fromCallTime, toReturnTime).map(
                            record -> {
                                Long callPathId = record.get(0, Long.class);
                                String name = record.get(1, String.class);
                                String shortName = record.get(2, String.class);
                                Long totalTime = record.get(3, Long.class);
                                Long totalInsnCount = record.get(4, Long.class);
                                Long totalCycCount = record.get(5, Long.class);
                                Long totalBranchCount = record.get(6, Long.class);
                                Long callTime = record.get(7, Long.class);
                                Long returnTime = record.get(8, Long.class);

                                return new CallTreeData(name, totalTime, callPathId, totalTime, callTime, returnTime);
                            }
                    );
            } else {
                return queryContext.fetch(
                    "SELECT call_path_id, name, short_name, return_time - call_time, " +
                            "insn_count, cyc_count, branch_count, call_time, return_time " +
                            "FROM calls " +
                            "INNER JOIN call_paths ON calls.call_path_id = call_paths.id " +
                            "INNER JOIN symbols ON call_paths.symbol_id = symbols.id " +
                            "INNER JOIN dsos ON symbols.dso_id = dsos.id " +
                            "WHERE parent_call_path_id = ? " +
                            "AND comm_id = ? " +
                            "AND thread_id = ? " +
                            "ORDER BY call_time, call_path_id",
                    parentCallPathId, commId, threadId).map(
                            record -> {
                                Long callPathId = record.get(0, Long.class);
                                String name = record.get(1, String.class);
                                String shortName = record.get(2, String.class);
                                Long totalTime = record.get(3, Long.class);
                                Long totalInsnCount = record.get(4, Long.class);
                                Long totalCycCount = record.get(5, Long.class);
                                Long totalBranchCount = record.get(6, Long.class);
                                Long callTime = record.get(7, Long.class);
                                Long returnTime = record.get(8, Long.class);

                                return new CallTreeData(name, totalTime, callPathId, totalTime, callTime, returnTime);
                            }
                    );
            }
        }
    }
} 

File: src/main/kotlin/com/perfscope/db/DatabaseLoader.kt
package com.perfscope.db

import com.perfscope.model.Call
import com.perfscope.model.Command
import com.perfscope.model.Thread
import com.perfscope.model.tables.Comms
import com.perfscope.model.tables.references.CALLS
import com.perfscope.model.tables.references.CALL_PATHS
import com.perfscope.model.tables.references.COMMS
import com.perfscope.model.tables.references.COMM_THREADS
import com.perfscope.model.tables.references.DSOS
import com.perfscope.model.tables.references.SYMBOLS
import com.perfscope.model.tables.references.THREADS
import org.jooq.DSLContext
import org.jooq.impl.DSL
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.sql.SQLException
import kotlin.time.Duration
import kotlin.time.DurationUnit
import kotlin.time.toDuration

class DatabaseLoader(private val dbPath: String) {

    companion object {
        val logger: Logger = LoggerFactory.getLogger(DatabaseLoader::class.java)
        const val ROOT_PARENT_CALL_PATH_ID: Long = 1L
    }

    @Throws(SQLException::class)
    fun fetchCommands(dbPath: String): List<Command> {
        val result = mutableListOf<Command>()

        DatabaseConnectionHolder(dbPath).use { connection: DatabaseConnectionHolder ->
            val context: DSLContext = connection.context
            val commandsWithCalls = context.selectFrom(Comms.COMMS)
                .where(COMMS.HAS_CALLS.eq(true))
                .fetch()

            commandsWithCalls.forEach { command ->
                val threads = context.select<Long?, Int?, Int?>(
                    COMM_THREADS.THREAD_ID,
                    THREADS.PID,
                    THREADS.TID)
                    .from(COMM_THREADS)
                    .innerJoin(THREADS).on(COMM_THREADS.THREAD_ID.eq(THREADS.ID.cast(Long::class.java)))
                    .where(COMM_THREADS.COMM_ID.eq(command.id?.toLong()))
                    .and(COMM_THREADS.THREAD_ID.isNotNull)
                    .fetch()
                    .map { (threadId, pid, tid) -> Thread(threadId, pid, tid) }
                val commandText = command.comm
                if (commandText != null) {
                    result += Command(command.id!!.toLong(), commandText, threads)
                }
            }

            logger.info("Found ${commandsWithCalls.size} commands with calls")
        }
        return result
    }

    fun fetchCalls(commandId: Long, threadId: Long, parentCallPathId: Long, fromCallTime: Long?, toReturnTime: Long?): List<Call> {
        DatabaseConnectionHolder(dbPath).use { connection: DatabaseConnectionHolder ->
            var select = connection.context.select(
                CALLS.CALL_PATH_ID,
                SYMBOLS.NAME,
                CALLS.CALL_TIME,
                CALLS.RETURN_TIME,
                CALLS.RETURN_TIME - CALLS.CALL_TIME)
                .from(CALLS)
                .innerJoin(CALL_PATHS).on(CALLS.CALL_PATH_ID.eq(CALL_PATHS.ID.cast(Long::class.java)))
                .innerJoin(SYMBOLS).on(CALL_PATHS.SYMBOL_ID.eq(SYMBOLS.ID.cast(Long::class.java)))
                .innerJoin(DSOS).on(SYMBOLS.DSO_ID.eq(DSOS.ID.cast(Long::class.java)))
                .where(CALLS.PARENT_CALL_PATH_ID.eq(parentCallPathId))
                .and(CALLS.COMM_ID.eq(commandId))
                .and(CALLS.THREAD_ID.eq(threadId))
            if (fromCallTime != null && toReturnTime != null) {
                select = select
                    .and(CALLS.CALL_TIME.greaterOrEqual(fromCallTime))
                    .and(CALLS.RETURN_TIME.lessOrEqual(toReturnTime))
            }

            val result = select.orderBy(CALLS.CALL_TIME, CALLS.CALL_PATH_ID)
                .fetch()
                .map { (callPathId, name, callTime, returnTime, duration) ->
                    Call(name, duration, callPathId, duration, callTime, returnTime)
                }
            return result
        }
    }

    fun search(commandId: Long, threadId: Long, searchTerm: String) : Boolean {
        DatabaseConnectionHolder(dbPath).use { connection: DatabaseConnectionHolder ->
            return connection.context.select(THREADS.TID)
                .from(CALLS)
                .innerJoin(THREADS).on(CALLS.THREAD_ID.eq(THREADS.ID.cast(Long::class.java)))
                .innerJoin(CALL_PATHS).on(CALLS.CALL_PATH_ID.eq(CALL_PATHS.ID.cast(Long::class.java)))
                .innerJoin(SYMBOLS).on(CALL_PATHS.SYMBOL_ID.eq(SYMBOLS.ID.cast(Long::class.java)))
                .where(CALLS.COMM_ID.eq(commandId))
                .and(CALLS.THREAD_ID.eq(threadId))
                .and(SYMBOLS.NAME.like("%${searchTerm}%"))
                .limit(1)
                .fetch()
                .map { (id) -> id }
                .isEmpty()
                .not()
        }
    }

    fun calcTotalTime(commandId: Long, threadId: Long): Duration {
        DatabaseConnectionHolder(dbPath).use { connection: DatabaseConnectionHolder ->
            val totalDurationNanos = connection.context.select(DSL.sum(CALLS.RETURN_TIME - CALLS.CALL_TIME).cast(Long::class.java))
                .from(CALLS)
                .where(CALLS.PARENT_CALL_PATH_ID.eq(ROOT_PARENT_CALL_PATH_ID))
                .and(CALLS.COMM_ID.eq(commandId))
                .and(CALLS.THREAD_ID.eq(threadId))
                .groupBy(CALLS.CALL_PATH_ID)
                .fetch()
                .map { (value) -> value }
                .sumOf { it }
            return totalDurationNanos.toDuration(DurationUnit.NANOSECONDS)
        }
    }
}